nft-client/src/main.rs

```diff
@@ -0,0 +1,635 @@
+            );
+            let hash =
+                send_transaction(&mut grpc_client, &account_data, transaction_payload).await?;
+            println!("Transaction send {}", hash);
+        }
+    }
+
+    Ok(())
+}
+
+/// Collect every contract event in a transaction summary for a given contract
+/// address.
+fn collect_summary_contract_events(
+    entry: DatabaseSummaryEntry,
+    contract: types::ContractAddress,
+) -> Vec<smart_contracts::ContractEvent> {
+    if let DatabaseSummaryEntry::BlockItem(item) = entry {
+        if let types::BlockItemSummaryDetails::AccountTransaction(tx) = item.details {
+            if let types::AccountTransactionEffects::ContractUpdateIssued { effects } = tx.effects {
+                let mut events = Vec::new();
+                for effect in effects {
+                    if let types::ContractTraceElement::Updated { data } = effect {
+                        if data.address == contract {
+                            events.extend(data.events)
+                        }
+                    }
+                }
+                return events;
+            } else if let types::AccountTransactionEffects::ContractInitialized { data } =
+                tx.effects
+            {
+                if data.address == contract {
+                    return data.events;
+                }
+            }
+        }
+    }
+    Vec::new()
+}
+
+/// Convert an account address from `types` to an account address from
+/// `concordium_contracts_common`.
+fn convert_account_address(
+    account: &id::types::AccountAddress,
+) -> concordium_contracts_common::AccountAddress {
+    let mut address_bytes = [0u8; 32];
+    address_bytes.copy_from_slice(account.as_ref());
+    concordium_contracts_common::AccountAddress(address_bytes)
+}
+
+/// Convert a contract address from `types` to a contract address from
+/// `concordium_contracts_common`.
+fn convert_contract_address(
+    contract: &types::ContractAddress,
+) -> concordium_contracts_common::ContractAddress {
+    concordium_contracts_common::ContractAddress {
+        index:    contract.index.index,
+        subindex: contract.subindex.sub_index,
+    }
+}
+
+/// Convert an address from `types` to an address from
+/// `concordium_contracts_common`.
+fn convert_address(address: types::Address) -> concordium_contracts_common::Address {
+    match address {
+        types::Address::Account(addr) => {
+            concordium_contracts_common::Address::Account(convert_account_address(&addr))
+        }
+        types::Address::Contract(addr) => {
+            concordium_contracts_common::Address::Contract(convert_contract_address(&addr))
+        }
+    }
+}
+
+/// Print the NFT contract state in a human readable format.
+fn pretty_print_contract_state(state: &NFTContractState) {
+    for (owner, address_state) in &state.0 {
+        if address_state.owned_tokens.is_empty() && address_state.operators.is_empty() {
+            continue;
+        }
```

Comment:
Why do we print nothing at all? Isn't that ambiguous?

Reply:
Well, the contract does not clean up the contract state, so if an account address at some point owned a token, but does not anymore, the address would still be in the contract state.
Ideally the contract would make sure to clean up.

Reply:
But would it not make sense to print something here so that it is clear this is the situation?

Reply:
I have removed this check